include/mbgl/gfx/drawable.hpp

```diff
@@ -0,0 +1,230 @@
+
+    /// Is custom
+    bool getIsCustom() const { return isCustom; }
+    void setIsCustom(bool value) { isCustom = value; }
+
+    /// The ID of the tile that this drawable represents, if any
+    const std::optional<OverscaledTileID>& getTileID() const { return tileID; }
+    void setTileID(const OverscaledTileID& value) { tileID = value; }
+
+    const gfx::CullFaceMode& getCullFaceMode() const;
+    void setCullFaceMode(const gfx::CullFaceMode&);
+
+    const gfx::ColorMode& getColorMode() const;
+    void setColorMode(const gfx::ColorMode&);
+
+    /// Get the vertex attributes that override default values in the shader program
+    virtual const gfx::VertexAttributeArray& getVertexAttributes() const = 0;
+    virtual gfx::VertexAttributeArray& mutableVertexAttributes() = 0;
+
+    virtual void setVertexAttributes(const gfx::VertexAttributeArray&) = 0;
+    virtual void setVertexAttributes(gfx::VertexAttributeArray&&) = 0;
+
+    /// Provide raw data for vertices
+    /// Incompatible with adding primitives
+    virtual void setVertices(std::vector<uint8_t>&&, std::size_t, AttributeDataType) = 0;
+
+    /// Provide raw indexes and segments
+    void setIndexData(std::vector<std::uint16_t> indexes, std::vector<UniqueDrawSegment>);
+    virtual void setIndexData(gfx::IndexVectorBasePtr, std::vector<UniqueDrawSegment>) = 0;
+
+    /// Get the tweakers attached to this drawable
+    const std::vector<DrawableTweakerPtr>& getTweakers() const { return tweakers; }
+    void addTweaker(DrawableTweakerPtr value) { tweakers.emplace_back(std::move(value)); }
+    void setTweakers(std::vector<DrawableTweakerPtr> value) { tweakers = std::move(value); }
+    void clearTweakers() { tweakers.clear(); }
+
+    /// Get the uniform buffers attached to this drawable
+    virtual const gfx::UniformBufferArray& getUniformBuffers() const = 0;
+    virtual gfx::UniformBufferArray& mutableUniformBuffers() = 0;
+
+    const UniqueDrawableData& getData() const { return drawableData; }
+    void setData(UniqueDrawableData&& value) { drawableData = std::move(value); }
+
+protected:
+    bool enabled = true;
+    bool enableColor = true;
+    bool enableStencil = false;
+    bool is3D = false;
+    bool isCustom = false;
+    std::string name;
+    const util::SimpleIdentity uniqueID;
+    gfx::ShaderProgramBasePtr shader;
+    mbgl::RenderPass renderPass;
+    std::optional<OverscaledTileID> tileID;
+    DrawPriority drawPriority = 0;
+    int32_t lineWidth = 1;
+    int32_t subLayerIndex = 0;
+    DepthMaskType depthType; // = DepthMaskType::ReadOnly;
+    UniqueDrawableData drawableData{};
+
+    struct Impl;
+    std::unique_ptr<Impl> impl;
+
+    Textures textures;
+    std::vector<DrawableTweakerPtr> tweakers;
+};
+
+using DrawablePtr = std::shared_ptr<Drawable>;
+using UniqueDrawable = std::unique_ptr<Drawable>;
+
+/// Comparator for sorting drawable pointers primarily by draw priority
+struct DrawableLessByPriority {
```

Comment:
Let's use this thread to discuss tests. If there are tests already they should be added to this PR. Without tests I fear a large part of the code base will become read only after the Metal project has concluded, as we will be too afraid to break anything.

At a minimum I think we should create the files for each new concept, so we can easily add more tests later. I think there is a lot of low-hanging fruit like this `struct`. What do you think?

Reply:
What sort of tests do you mean?

Reply:
Ah, class level tests.  It's worth having that discussion, but I've never been much of a fan since they got popular in Java.  It's a different style of working, so if someone wants to push for it, I'll listen.

Reply:
There are a lot of tests under the ./test directory. and you can run it via mbgl-test project.

Reply:
@alanchenboy in TomTom we have per test case tweakable thresholds and we do use exactly same images for OpenGL and Metal API. The difference indeed usually is on the edges of the geometry where opacity of the pixels don't match sometimes.

There is a possible way of working around it. In OpenCV there are algos for edge detection and then some pixels around the edges can have increased tolerance for pixel-to-pixel comparison.

Reply:
This looks like a great solution, cause some shared use line primitive in OpenGL, it's maybe a little bit hard to make it looks same in Metal, ignoring these edges can make unit tests more neat.

Reply:
Yes my comment was relating to extending the unit tests (in `./test`) specifically.  What @mwilsnd says is in line with my thinking.

Unit tests are very helpful to keep things functional as other contributors start poking around in the codebase. In addition, they also serve as documentation. For example, a test of VertexAttributeArray::readDataDrivenPaintProperties would be helpful to understand how it is used. Good test coverage also makes my life as maintainer much easier. I don't need to do a deep code dive of very PR to ensure correctness of the requested changes. 

I'm not saying we should add a test for every getter and setter, Mapbox also took unit tests (too?) far, but if you ask me it will help the library a lot if all non-trivial logic has a few tests.

Reply:
Yes my comment was relating to extending the unit tests (in ./test) specifically. What @mwilsnd says is in line with my thinking.

Unit tests are very helpful to keep things functional as other contributors start poking around in the codebase. In addition, they also serve as documentation. For example, a test of VertexAttributeArray::readDataDrivenPaintProperties would be helpful to understand how it is used. Good test coverage also makes my life as maintainer much easier. I don't need to do a deep code dive of very PR to ensure correctness of the requested changes.

I'm not saying we should add a test for every getter and setter, Mapbox also took unit tests (too?) far, but if you ask me it will help the library a lot if all non-trivial logic has a few tests.

Reply:
Render tests are going to be more complicated but I can see value in at least attempting to reuse the same expected images from the OpenGL metrics. I want to preserve the probe functionality at a minimum and generate new probe metric expectations for the Metal renderer, but I can also see generating a fresh set of expected images for the new pathway as the old ones reach a state of acceptable correctness.

Reply:
For scheduling, let's bring in the render tests after we've got things looking more or less right in Metal.  The probe tests let's schedule near the end of phase 2.